consensus/src/chained_bft/consensus_types/block_test.rs

```diff
@@ -190,6 +196,48 @@ fn test_genesis() {
     assert!(genesis_block.is_genesis_block());
 }
 
+#[test]
+fn test_nil_block() {
+    let genesis_block = Block::make_genesis_block();
+    let quorum_cert = QuorumCert::certificate_for_genesis();
+
+    let nil_block = Block::make_nil_block(&genesis_block, 1, quorum_cert);
+    assert_eq!(
+        nil_block.quorum_cert().certified_block_id(),
+        genesis_block.id()
+    );
+    assert_eq!(nil_block.round(), 1);
+    assert_eq!(nil_block.timestamp_usecs(), genesis_block.timestamp_usecs());
+    assert_eq!(nil_block.is_nil_block(), true);
+    assert!(nil_block.author().is_none());
+
+    let dummy_verifier = Arc::new(ValidatorVerifier::<Ed25519PublicKey>::new(HashMap::new()));
+    assert!(nil_block.verify(dummy_verifier.as_ref()).is_ok());
```

Comment:
this doesn't verify anything?

Reply:
It verifies that a nil block passes verification successfully :). Nil block verification doesn't check the author signature of the id, but verifies all the other stuff.

Reply:
the function only checks the signature and QC's signature, both need a non-dummy verifier.
I'm surprised we don't check self.hash() == self.id() in verify()

Reply:
In this test the NIL block is keeping a QC to genesis, which has its own special case in `QuorumCert::verify()` that doesn't really check signatures (cause there are none), hence dummy verifier is enough.

However, I think you just found a bug in the `verify()` function that has nothing to do with the NIL blocks! We indeed do not verify that the id of the block corresponds to its hash. 
An attacker could exploit it by assigning the ids of some previously committed blocks to the newly generated proposals. Let me fix that.